File: CSharpCodeBase/components/controlsystem/enemyactions/actiondashpunch.lua
local Action = require 'components.controlsystem.action'
local Class = require 'utils.hump.class'

local ActionDashPunch = Class({__includes=Action, init = function (self, entity, target)
      Action.init(self, 4, false, 'dashpunch')
      self.entity = entity
      self.target = target
end})

function ActionDashPunch:OnStartRunning()
  self.entity.mover:Stop()
  self.entity.combat:Attack(self.target)
  self.isAnimationStarted = false
  self.jumping = false
  self.canpunch = false
  self.dv = self.target:GetPosition() - self.entity:GetPosition()
  self.dv:Normalize()
end

function ActionDashPunch:Update()
  if not self.isAnimationStarted then
    self.isAnimationStarted = self.entity.mecanim:CheckStateName('Action')
  end
  return self.isAnimationStarted and not self.entity.mecanim:CheckStateName("Action")
end

function ActionDashPunch:FixedUpdate()
  if not self.jumping then
    self.entity.mover:LookAt(self.target:GetPosition())
  end
  local speedCurve = self.entity.mecanim:GetFloat('speed_curve')
  --local dv = self.target:GetPosition() - self.entity:GetPosition()
  --dv:Normalize()
  if self.jumping then
    RigidbodyUtils.MoveNotRotate(self.entity.rigidbody, self.dv * speedCurve * 10)
  end
  if self.canpunch then
    if self.target.interactable and self.entity.weaponContainer:CanAttack(self.target) then
      self.entity.weaponContainer:Attack(self.target)
      --self.entity.damageProcessor:SendDamage(self.target, self.entity.weaponContainer.currentWeapon:GetDamage())
      self.canpunch = false
    end
  end
end


function ActionDashPunch:OnEvent(data)
  if data == "punch" then        
    --if self.target.interactable and self.entity.weaponContainer:CanAttack(self.target) then
    --  self.entity.damageProcessor:SendDamage(self.target, self.entity.weaponContainer.currentWeapon:GetDamage())
    --end
  elseif data == "jumping" then
    self.jumping = true
    self.canpunch = true
    self.dv = self.target:GetPosition() - self.entity:GetPosition()
    self.dv:Normalize()    
  elseif data == "stoppunch" then
    self.canpunch = false
  end
  
end

function ActionDashPunch:OnStopRunning()
  self.entity.mover:Stop()
end

return ActionDashPunch

File: CSharpCodeBase/components/controlsystem/enemyactions/actiondashpunch.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class actiondashpunch :action{
 public void init (self, entity, target);
       Action.init(self, 4, false, "dashpunch");
       this.entity = entity;
       this.target = target;
 }})
 public void OnStartRunning(){
   this.entity.mover:Stop();
   this.entity.combat:Attack(this.target);
   this.isAnimationStarted = false;
   this.jumping = false;
   this.canpunch = false;
   this.dv = this.target:GetPosition() - this.entity:GetPosition();
   this.dv:Normalize();
 }
 public void Update(){
   if(not this.isAnimationStarted  ){ 
     this.isAnimationStarted = this.entity.mecanim:CheckStateName("Action");
   }
   return this.isAnimationStarted  &&  not this.entity.mecanim:CheckStateName("Action");
 }
 public void FixedUpdate(){
   if(not this.jumping  ){ 
     this.entity.mover:LookAt(this.target:GetPosition());
   }
   var speedCurve = this.entity.mecanim:GetFloat("speed_curve");
   //local dv = this.target:GetPosition() - this.entity:GetPosition()
   //dv:Normalize()
   if(this.jumping  ){ 
     RigidbodyUtils.MoveNotRotate(this.entity.rigidbody, this.dv * speedCurve * 10);
   }
   if(this.canpunch  ){ 
     if(this.target.interactable  &&  this.entity.weaponContainer:CanAttack(this.target)  ){ 
       this.entity.weaponContainer:Attack(this.target);
       //this.entity.damageProcessor:SendDamage(this.target, this.entity.weaponContainer.currentWeapon:GetDamage())
       this.canpunch = false;
     }
   }
 }
 public void OnEvent(data){
   if(data == "punch"  ){         
     //if this.target.interactable  &&  this.entity.weaponContainer:CanAttack(this.target)  ){ 
     //  this.entity.damageProcessor:SendDamage(this.target, this.entity.weaponContainer.currentWeapon:GetDamage())
     //end
   }else{if data == "jumping"  ){ 
     this.jumping = true;
     this.canpunch = true;
     this.dv = this.target:GetPosition() - this.entity:GetPosition();
     this.dv:Normalize()    ;
   }else{if data == "stoppunch"  ){ 
     this.canpunch = false;
   }
   
 }
 public void OnStopRunning(){
   this.entity.mover:Stop();
 }
 }}